File: packages/server/cmd/renewlet/media_icon_index_test.go
package main

// 图标索引 route 测试保护管理员边界、空 body 契约、失败不切 active，以及媒体候选切换 runtime resolver。

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestMediaIconIndexStatusRequiresAdmin(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, userToken := createRouteTestUser(t, app, "user")
	_, adminToken := createRouteTestUser(t, app, "admin")

	cases := []struct {
		name     string
		token    string
		wantCode int
	}{
		{name: "anonymous", token: "", wantCode: http.StatusUnauthorized},
		{name: "non admin", token: userToken, wantCode: http.StatusForbidden},
		{name: "admin", token: adminToken, wantCode: http.StatusOK},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			res := serveTestRequest(t, app, http.MethodGet, "/api/app/admin/media/icon-index", "", tc.token)
			if res.Code != tc.wantCode {
				t.Fatalf("expected status %d, got %d: %s", tc.wantCode, res.Code, res.Body.String())
			}
		})
	}
}

func TestMediaIconIndexRefreshUsesStrictEmptyBody(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, adminToken := createRouteTestUser(t, app, "admin")

	res := serveTestRequest(t, app, http.MethodPost, "/api/app/admin/media/icon-index/providers/thesvg/refresh", `{}`, adminToken)
	if res.Code != http.StatusBadRequest {
		t.Fatalf("expected refresh with JSON body to return 400, got %d: %s", res.Code, res.Body.String())
	}
}

func TestMediaIconIndexStatusUsesSeedProviderVersions(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, adminToken := createRouteTestUser(t, app, "admin")

	res := serveTestRequest(t, app, http.MethodGet, "/api/app/admin/media/icon-index", "", adminToken)
	if res.Code != http.StatusOK {
		t.Fatalf("expected status 200, got %d: %s", res.Code, res.Body.String())
	}
	var status builtInIconIndexStatusResponse
	if err := json.NewDecoder(res.Body).Decode(&status); err != nil {
		t.Fatal(err)
	}
	assertSeedProviderVersions(t, status, nil)
}

func TestMediaIconIndexCheckProviderDoesNotSwitchActive(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, adminToken := createRouteTestUser(t, app, "admin")

	server := newMediaIconRegistryServer(t, http.StatusOK)
	withBuiltInIconGitHubAPIBase(t, server.URL)
	res := serveTestRequest(t, app, http.MethodPost, "/api/app/admin/media/icon-index/providers/thesvg/check", "", adminToken)
	if res.Code != http.StatusOK {
		t.Fatalf("expected check success 200, got %d: %s", res.Code, res.Body.String())
	}
	var response builtInIconIndexProviderCheckResponse
	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
		t.Fatal(err)
	}
	if response.Status.Source != "embedded" || response.Status.Refreshing || response.Provider.Refreshing || response.Provider.Latest == nil || response.Provider.Latest.CommitSHA == nil {
		t.Fatalf("unexpected check status: %#v", response.Status)
	}
	record, err := findMediaIconIndexRecord(app)
	if err != nil {
		t.Fatal(err)
	}
	if record == nil || record.GetString("hash") != "" || record.GetString("indexGzipBase64") != "" {
		t.Fatalf("expected check to keep active index empty, got %#v", record)
	}
}

func TestMediaIconIndexCheckProviderRecordsGitHubRateLimitWithoutSwitchingActive(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, adminToken := createRouteTestUser(t, app, "admin")

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("X-RateLimit-Remaining", "0")
		w.Header().Set("X-RateLimit-Reset", "1781190000")
		http.Error(w, "rate limited", http.StatusForbidden)
	}))
	defer server.Close()
	withBuiltInIconGitHubAPIBase(t, server.URL)

	res := serveTestRequest(t, app, http.MethodPost, "/api/app/admin/media/icon-index/providers/selfhst/check", "", adminToken)
	if res.Code != http.StatusOK {
		t.Fatalf("expected rate limited check to keep a usable 200 response, got %d: %s", res.Code, res.Body.String())
	}
	var response builtInIconIndexProviderCheckResponse
	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
		t.Fatal(err)
	}
	if response.Provider.LastError == nil || !strings.Contains(*response.Provider.LastError, "RENEWLET_GITHUB_TOKEN") {
		t.Fatalf("expected rate limit guidance in lastError, got %#v", response.Provider.LastError)
	}
	record, err := findMediaIconIndexRecord(app)
	if err != nil {
		t.Fatal(err)
	}
	if record == nil || record.GetString("hash") != "" || record.GetString("indexGzipBase64") != "" {
		t.Fatalf("expected failed check to keep active index empty, got %#v", record)
	}
}

func TestMediaIconIndexGitHubRequestUsesTokenAndUserAgent(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, adminToken := createRouteTestUser(t, app, "admin")
	t.Setenv("RENEWLET_GITHUB_TOKEN", "github-token")

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Authorization") != "Bearer github-token" {
			t.Fatalf("expected GitHub token authorization header, got %q", r.Header.Get("Authorization"))
		}
		if r.Header.Get("User-Agent") == "" {
			t.Fatal("expected GitHub request user agent")
		}
		w.Header().Set("content-type", "application/json")
		switch r.URL.Path {
		case "/repos/selfhst/icons/commits/main":
			_, _ = w.Write([]byte(`{"sha":"abc1234567890abcdef","commit":{"committer":{"date":"2026-06-11T00:00:00Z"}}}`))
		default:
			http.NotFound(w, r)
		}
	}))
	defer server.Close()
	withBuiltInIconGitHubAPIBase(t, server.URL)

	res := serveTestRequest(t, app, http.MethodPost, "/api/app/admin/media/icon-index/providers/selfhst/check", "", adminToken)
	if res.Code != http.StatusOK {
		t.Fatalf("expected check success 200, got %d: %s", res.Code, res.Body.String())
	}
}

func TestMediaIconIndexRefreshSuccessAndFailureKeepsPreviousRuntimeIndex(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, adminToken := createRouteTestUser(t, app, "admin")

	successServer := newMediaIconRegistryServer(t, http.StatusOK)
	withMediaResolverProviderBase(t, successServer.URL)
	withBuiltInIconGitHubAPIBase(t, successServer.URL)
	res := serveTestRequest(t, app, http.MethodPost, "/api/app/admin/media/icon-index/providers/thesvg/refresh", "", adminToken)
	if res.Code != http.StatusOK {
		t.Fatalf("expected refresh success 200, got %d: %s", res.Code, res.Body.String())
	}
	var success builtInIconIndexProviderRefreshResponse
	if err := json.NewDecoder(res.Body).Decode(&success); err != nil {
		t.Fatal(err)
	}
	if success.Status.Source != "runtime" || success.Status.Refreshing || success.Provider.Refreshing || success.Status.ProviderCounts.TheSVG != 1 || success.Status.Hash == nil || success.Provider.Current == nil || success.Provider.Current.CommitSHA == nil {
		t.Fatalf("unexpected success status: %#v", success.Status)
	}
	assertSeedProviderVersions(t, success.Status, map[string]bool{"thesvg": true})
	activeHash := *success.Status.Hash

	failingServer := newMediaIconRegistryServer(t, http.StatusInternalServerError)
	withMediaResolverProviderBase(t, failingServer.URL)
	withBuiltInIconGitHubAPIBase(t, failingServer.URL)
	res = serveTestRequest(t, app, http.MethodPost, "/api/app/admin/media/icon-index/providers/thesvg/refresh", "", adminToken)
	if res.Code != http.StatusBadGateway {
		t.Fatalf("expected refresh failure 502, got %d: %s", res.Code, res.Body.String())
	}
	var failure builtInIconIndexProviderRefreshResponse
	if err := json.NewDecoder(res.Body).Decode(&failure); err != nil {
		t.Fatal(err)
	}
	if failure.Status.Source != "runtime" || failure.Status.Refreshing || failure.Provider.Refreshing || failure.Status.Hash == nil || *failure.Status.Hash != activeHash {
		t.Fatalf("expected failure to keep old active hash %q, got %#v", activeHash, failure.Status)
	}
	if failure.Provider.LastError == nil || *failure.Provider.LastError == "" {
		t.Fatalf("expected failure status to expose last error, got %#v", failure.Status)
	}
	if failure.Provider.Current == nil || failure.Provider.Current.CommitSHA == nil || *failure.Provider.Current.CommitSHA != *success.Provider.Current.CommitSHA {
		t.Fatalf("expected failed refresh to keep previous current version, got %#v", failure.Provider.Current)
	}
}

func TestMediaIconIndexRefreshReportsConcurrentLock(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, adminToken := createRouteTestUser(t, app, "admin")

	builtInIconIndexRefreshMu.Lock()
	defer builtInIconIndexRefreshMu.Unlock()

	res := serveTestRequest(t, app, http.MethodPost, "/api/app/admin/media/icon-index/providers/thesvg/refresh", "", adminToken)
	if res.Code != http.StatusConflict {
		t.Fatalf("expected locked refresh 409, got %d: %s", res.Code, res.Body.String())
	}
	var response builtInIconIndexProviderRefreshResponse
	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
		t.Fatal(err)
	}
	if !response.Status.Refreshing || !response.Provider.Refreshing {
		t.Fatalf("expected locked status to report refreshing: %#v", response.Status)
	}
}

func TestMediaCandidatesUseRuntimeIconIndex(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	_, token := createRouteTestUser(t, app, "user")

	icons := []builtInIcon{iconRecord(builtInIcon{
		Provider: "thesvg",
		Slug:     "runtime-only",
		Title:    "Runtime Only",
		Variants: []builtInIconVariant{{Name: "default", Path: "/public/icons/runtime-only/default.svg"}},
	})}
	hash, gzipBase64, err := encodeBuiltInIconIndex(icons)
	if err != nil {
		t.Fatal(err)
	}
	commit := "abc1234567890abcdef"
	short := "abc1234"
	version := &builtInIconProviderVersionResponse{
		SourceRef:          commit,
		DisplayVersion:     short,
		CommitSHA:          &commit,
		CommitShortSHA:     &short,
		CommitDate:         nil,
		ReleaseTag:         nil,
		ReleasePublishedAt: nil,
	}
	if err := saveMediaIconProviderRefreshSuccess(app, "thesvg", "2026-06-11T00:00:00Z", hash, gzipBase64, icons, version, ""); err != nil {
		t.Fatal(err)
	}

	res := serveTestRequest(t, app, http.MethodPost, "/api/app/media/candidates", `{"kind":"logo","mode":"search","items":[{"id":"runtime","name":"Runtime Only"}],"limit":4}`, token)
	if res.Code != http.StatusOK {
		t.Fatalf("expected media candidates 200, got %d: %s", res.Code, res.Body.String())
	}
	var response mediaCandidateResolveResponse
	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
		t.Fatal(err)
	}
	if len(response.Items) != 1 || len(response.Items[0].Candidates.BuiltIn) == 0 {
		t.Fatalf("expected runtime built-in candidate, got %#v", response)
	}
	candidate := response.Items[0].Candidates.BuiltIn[0]
	if candidate.Label != "Runtime Only" || candidate.Provider != "thesvg" || candidate.MatchedQuery != "runtime only" {
		t.Fatalf("unexpected runtime candidate: %#v", candidate)
	}
}

func assertSeedProviderVersions(t *testing.T, status builtInIconIndexStatusResponse, skip map[string]bool) {
	t.Helper()
	for _, provider := range status.Providers {
		if skip != nil && skip[provider.Provider] {
			continue
		}
		if provider.Current == nil || provider.Current.CommitSHA == nil || provider.Current.CommitShortSHA == nil || provider.Current.CommitDate == nil {
			t.Fatalf("expected %s to expose seed commit metadata, got %#v", provider.Provider, provider.Current)
		}
		if provider.Current.SourceRef == "embedded" || provider.Current.SourceRef == "runtime" {
			t.Fatalf("expected %s current version to use a real commit sourceRef, got %#v", provider.Provider, provider.Current)
		}
		if len(*provider.Current.CommitShortSHA) != 7 {
			t.Fatalf("expected %s short sha length 7, got %q", provider.Provider, *provider.Current.CommitShortSHA)
		}
	}
}

func newMediaIconRegistryServer(t *testing.T, statusCode int) *httptest.Server {
	t.Helper()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if statusCode != http.StatusOK {
			http.Error(w, "registry failed", statusCode)
			return
		}
		w.Header().Set("content-type", "application/json")
		switch r.URL.Path {
		case "/repos/glincker/thesvg/commits/main", "/repos/selfhst/icons/commits/main", "/repos/homarr-labs/dashboard-icons/commits/main":
			_, _ = w.Write([]byte(`{"sha":"abc1234567890abcdef","commit":{"committer":{"date":"2026-06-11T00:00:00Z"}}}`))
		case "/repos/glincker/thesvg/releases/latest":
			_, _ = w.Write([]byte(`{"tag_name":"thesvg@9.9.9","published_at":"2026-06-11T00:00:00Z"}`))
		case "/src/data/icons.json":
			_, _ = w.Write([]byte(`[{"slug":"go-registry","title":"Go Registry","variants":{"default":"/icons/go-registry/default.svg"}}]`))
		case "/index.json":
			_, _ = w.Write([]byte(`[{"Reference":"go-selfhst","Name":"Go selfh.st","SVG":"Yes","Light":"N","Dark":"N"}]`))
		case "/metadata.json":
			_, _ = w.Write([]byte(`{"go-dashboard":{"aliases":["Go Dashboard"],"categories":["Test"]}}`))
		case "/tree.json":
			_, _ = w.Write([]byte(`{"svg":["go-dashboard.svg"]}`))
		default:
			http.NotFound(w, r)
		}
	}))
	t.Cleanup(server.Close)
	return server
}

func withMediaResolverProviderBase(t *testing.T, base string) {
	t.Helper()
	previous := mediaResolverCfg.BuiltInProviders
	next := make([]struct {
		Provider string `json:"provider"`
		CDNBase  string `json:"cdnBase"`
		GitHub   struct {
			Owner         string `json:"owner"`
			Repo          string `json:"repo"`
			Branch        string `json:"branch"`
			LatestRelease bool   `json:"latestRelease"`
		} `json:"github"`
		PreferredVariants []string `json:"preferredVariants"`
	}, len(previous))
	copy(next, previous)
	for index := range next {
		next[index].CDNBase = base
	}
	mediaResolverCfg.BuiltInProviders = next
	t.Cleanup(func() {
		mediaResolverCfg.BuiltInProviders = previous
	})
}

func withBuiltInIconGitHubAPIBase(t *testing.T, base string) {
	t.Helper()
	previous := builtInIconGitHubAPIBase
	builtInIconGitHubAPIBase = base
	t.Cleanup(func() {
		builtInIconGitHubAPIBase = previous
	})
}

File: packages/cloudflare/src/media-icon-index.test.ts
// Worker 图标索引测试保护 D1 metadata、R2 gzip blob、provider 刷新锁和失败不切 active 的运行面契约。
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { resolveMediaCandidateItem } from "@renewlet/shared/media-resolver";
import type { BuiltInIconIndexProviderRefreshResponse, BuiltInIconIndexStatus } from "@renewlet/shared/schemas/media";
import {
  builtInIconIndexStatus,
  checkBuiltInIconIndexProvider,
  getActiveBuiltInMediaResolver,
  refreshBuiltInIconIndexProvider,
} from "./media-icon-index";
import type { Env, MediaIconIndexRow } from "./types";

vi.mock("./auth", () => ({
  requireAdmin: vi.fn().mockResolvedValue({
    user: { id: "usr_admin", role: "admin" },
  }),
}));

describe("Cloudflare media icon index", () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date("2026-06-11T00:00:00.000Z"));
  });

  afterEach(() => {
    vi.useRealTimers();
    vi.unstubAllGlobals();
  });

  it("returns embedded provider statuses when no runtime index is active", async () => {
    const env = createEnv();
    const response = await builtInIconIndexStatus(requestFixture("GET"), env);

    expect(response.status).toBe(200);
    const body = await response.json() as BuiltInIconIndexStatus;
    expect(body).toMatchObject({ source: "embedded", refreshing: false });
    expectSeedProviderVersions(body);
  });

  it("checks one provider latest version with ETag without switching active", async () => {
    const env = createEnv(mediaIconIndexRow({
      provider_status_json: JSON.stringify({
        thesvg: {
          latest: providerVersion("abc1234567890abcdef"),
          etag: "\"cached\"",
        },
      }),
    }));
    env.RENEWLET_GITHUB_TOKEN = "github-token";
    env.RENEWLET_VERSION = "1.2.3";
    const fetchMock = vi.fn(async () => new Response(null, { status: 304, headers: { etag: "\"cached\"" } }));
    vi.stubGlobal("fetch", fetchMock);

    const response = await checkBuiltInIconIndexProvider(requestFixture("POST"), env, "thesvg");

    expect(response.status).toBe(200);
    await expect(response.json()).resolves.toMatchObject({
      status: {
        refreshing: false,
      },
      provider: {
        provider: "thesvg",
        refreshing: false,
        latest: { commitSha: "abc1234567890abcdef" },
      },
    });
    expect(env.testState.row?.hash).toBeNull();
    expect(env.testState.objects.size).toBe(0);
    const calls = fetchMock.mock.calls as unknown as Array<[RequestInfo | URL, RequestInit?]>;
    const init = calls[0]?.[1];
    expect(init?.headers).toMatchObject({
      authorization: "Bearer github-token",
      "if-none-match": "\"cached\"",
      "user-agent": "Renewlet/1.2.3",
    });
  });

  it("records GitHub rate limits as provider status without breaking check responses", async () => {
    const env = createEnv();
    vi.stubGlobal("fetch", vi.fn(async () => new Response("rate limited", {
      status: 403,
      headers: {
        "x-ratelimit-remaining": "0",
        "x-ratelimit-reset": "1781190000",
      },
    })));

    const response = await checkBuiltInIconIndexProvider(requestFixture("POST"), env, "selfhst");

    expect(response.status).toBe(200);
    await expect(response.json()).resolves.toMatchObject({
      status: {
        source: "embedded",
        refreshing: false,
      },
      provider: {
        provider: "selfhst",
        refreshing: false,
        lastError: expect.stringContaining("RENEWLET_GITHUB_TOKEN"),
      },
    });
    expect(env.testState.row?.hash).toBeNull();
    expect(env.testState.objects.size).toBe(0);
  });

  it("refreshes one provider metadata and serves the runtime resolver from R2", async () => {
    const env = createEnv();
    stubRegistryFetch(registryFixtures());

    const response = await refreshBuiltInIconIndexProvider(requestFixture("POST"), env, "thesvg");

    expect(response.status).toBe(200);
    const body = await response.json() as BuiltInIconIndexProviderRefreshResponse;
    expect(body).toMatchObject({
      status: {
        source: "runtime",
        providerCounts: { thesvg: 1 },
        refreshing: false,
      },
      provider: {
        provider: "thesvg",
        iconCount: 1,
        refreshing: false,
        current: { commitSha: "abc1234567890abcdef" },
      },
    });
    expectSeedProviderVersions(body.status as BuiltInIconIndexStatus, ["thesvg"]);
    expect(env.testState.row?.r2_key).toMatch(/^system\/media-icon-index\/[a-f0-9]{64}\.json\.gz$/);
    expect(env.testState.objects.size).toBe(1);

    const resolver = await getActiveBuiltInMediaResolver(env);
    const item = resolveMediaCandidateItem(
      resolver,
      "logo",
      "search",
      { id: "cf", name: "CF Registry" },
      4,
      {},
    );
    expect(item.candidates.builtIn[0]?.label).toBe("CF Registry");
  });

  it("rejects non-empty refresh bodies before acquiring the lock", async () => {
    const env = createEnv();

    await expect(refreshBuiltInIconIndexProvider(requestFixture("POST", "{}"), env, "thesvg")).rejects.toMatchObject({
      status: 400,
      code: "NON_EMPTY_BODY",
    });
    expect(env.testState.row).toBeNull();
  });

  it("keeps the previous active index when a provider refresh fails", async () => {
    const env = createEnv();
    stubRegistryFetch(registryFixtures());
    const success = await refreshBuiltInIconIndexProvider(requestFixture("POST"), env, "thesvg");
    expect(success.status).toBe(200);
    const activeHash = env.testState.row?.hash;

    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response("failed", { status: 500 })));
    const failure = await refreshBuiltInIconIndexProvider(requestFixture("POST"), env, "thesvg");

    expect(failure.status).toBe(502);
    await expect(failure.json()).resolves.toMatchObject({
      status: {
        source: "runtime",
        hash: activeHash,
        refreshing: false,
      },
      provider: {
        provider: "thesvg",
        refreshing: false,
        lastError: expect.stringContaining("HTTP 500"),
      },
    });
    expect(env.testState.row?.hash).toBe(activeHash);
    expect(env.testState.row?.provider_status_json).toContain("HTTP 500");
  });

  it("reports a locked provider refresh without changing active metadata", async () => {
    const env = createEnv(mediaIconIndexRow({
      locked_until: "2026-06-11T00:01:00.000Z",
    }));

    const response = await refreshBuiltInIconIndexProvider(requestFixture("POST"), env, "thesvg");

    expect(response.status).toBe(409);
    await expect(response.json()).resolves.toMatchObject({
      status: {
        refreshing: true,
      },
      provider: {
        provider: "thesvg",
        refreshing: true,
      },
    });
    expect(env.testState.objects.size).toBe(0);
  });
});

function expectSeedProviderVersions(status: BuiltInIconIndexStatus, skipProviders: string[] = []): void {
  for (const provider of status.providers) {
    if (skipProviders.includes(provider.provider)) continue;
    expect(provider.current?.commitSha).toMatch(/^[a-f0-9]{40}$/);
    expect(provider.current?.commitShortSha).toMatch(/^[a-f0-9]{7}$/);
    expect(provider.current?.commitDate).toMatch(/^\d{4}-\d{2}-\d{2}T/);
    expect(provider.current?.sourceRef).not.toBe("embedded");
    expect(provider.current?.sourceRef).not.toBe("runtime");
  }
}

interface TestState {
  row: MediaIconIndexRow | null;
  objects: Map<string, Uint8Array>;
}

type TestEnv = Env & { testState: TestState };

function createEnv(row: MediaIconIndexRow | null = null): TestEnv {
  const state: TestState = {
    row,
    objects: new Map(),
  };

  return {
    DB: createD1(state),
    ASSETS_BUCKET: createR2(state),
    testState: state,
  };
}

function createD1(state: TestState): D1Database {
  return {
    prepare: (sql: string) => ({
      bind: (...args: unknown[]) => ({
        first: async () => {
          if (sql.includes("SELECT * FROM media_icon_indexes")) return state.row;
          return null;
        },
        run: async () => {
          if (sql.includes("INSERT OR IGNORE INTO media_icon_indexes")) {
            if (!state.row) {
              state.row = mediaIconIndexRow({
                provider_counts_json: "{}",
                provider_status_json: "{}",
                created_at: stringArg(args, 1),
                updated_at: stringArg(args, 2),
              });
              return d1Result(1);
            }
            return d1Result(0);
          }
          if (sql.includes("SET locked_until = ?")) {
            const now = stringArg(args, 3);
            if (!state.row?.locked_until || state.row.locked_until <= now) {
              state.row = { ...mediaIconIndexRow(state.row ?? {}), locked_until: stringArg(args, 0), updated_at: stringArg(args, 1) };
              return d1Result(1);
            }
            return d1Result(0);
          }
          if (sql.includes("SET locked_until = NULL")) {
            if (state.row) state.row = { ...state.row, locked_until: null, updated_at: stringArg(args, 0) };
            return d1Result(1);
          }
          if (sql.includes("SET hash = ?")) {
            state.row = {
              ...mediaIconIndexRow(state.row ?? {}),
              hash: stringArg(args, 0),
              r2_key: stringArg(args, 1),
              icon_count: numberArg(args, 2),
              provider_counts_json: stringArg(args, 3),
              provider_status_json: stringArg(args, 4),
              checked_at: stringArg(args, 5),
              index_updated_at: stringArg(args, 6),
              updated_at: stringArg(args, 7),
            };
            return d1Result(1);
          }
          if (sql.includes("SET checked_at = ?, provider_status_json = ?")) {
            state.row = {
              ...mediaIconIndexRow(state.row ?? {}),
              checked_at: stringArg(args, 0),
              provider_status_json: stringArg(args, 1),
              updated_at: stringArg(args, 2),
            };
            return d1Result(1);
          }
          return d1Result(0);
        },
      }),
    }),
  } as unknown as D1Database;
}

function createR2(state: TestState): R2Bucket {
  return {
    get: async (key: string) => {
      const bytes = state.objects.get(key);
      if (!bytes) return null;
      return {
        arrayBuffer: async () => bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength),
      };
    },
    put: async (key: string, value: ReadableStream | ArrayBuffer | ArrayBufferView | string | null | Blob) => {
      state.objects.set(key, await bodyToBytes(value));
      return null;
    },
  } as unknown as R2Bucket;
}

async function bodyToBytes(value: ReadableStream | ArrayBuffer | ArrayBufferView | string | null | Blob): Promise<Uint8Array> {
  if (value === null) return new Uint8Array();
  if (typeof value === "string") return new TextEncoder().encode(value);
  if (value instanceof ArrayBuffer) return new Uint8Array(value.slice(0));
  if (ArrayBuffer.isView(value)) {
    const buffer = new ArrayBuffer(value.byteLength);
    new Uint8Array(buffer).set(new Uint8Array(value.buffer, value.byteOffset, value.byteLength));
    return new Uint8Array(buffer);
  }
  if (value instanceof Blob) return new Uint8Array(await value.arrayBuffer());
  return new Uint8Array(await new Response(value).arrayBuffer());
}

function d1Result(changes: number): D1Result {
  return { meta: { changes } } as D1Result;
}

function mediaIconIndexRow(overrides: Partial<MediaIconIndexRow>): MediaIconIndexRow {
  return {
    key: "active",
    hash: null,
    r2_key: null,
    icon_count: 0,
    provider_counts_json: "{}",
    provider_status_json: "{}",
    checked_at: null,
    index_updated_at: null,
    locked_until: null,
    created_at: "2026-06-11T00:00:00.000Z",
    updated_at: "2026-06-11T00:00:00.000Z",
    ...overrides,
  };
}

function requestFixture(method: string, body?: string): Request {
  const headers: HeadersInit = {
    "accept-language": "en-US",
    authorization: "Bearer session-token",
  };
  const init: RequestInit = {
    method,
    headers,
  };
  if (body !== undefined) {
    headers["content-type"] = "application/json";
    init.body = body;
  }
  return new Request("https://renewlet.example/api/app/admin/media/icon-index/providers/thesvg/refresh", init);
}

function stubRegistryFetch(registries: Record<string, unknown>) {
  vi.stubGlobal("fetch", vi.fn(async (input: RequestInfo | URL) => {
    const url = String(input);
    if (url.includes("/repos/glincker/thesvg/commits/main")) {
      return jsonResponse({
        sha: "abc1234567890abcdef",
        commit: { committer: { date: "2026-06-11T00:00:00Z" } },
      }, { etag: "\"commit-etag\"" });
    }
    if (url.includes("/repos/glincker/thesvg/releases/latest")) {
      return jsonResponse({
        tag_name: "thesvg@9.9.9",
        published_at: "2026-06-11T00:00:00Z",
      });
    }
    if (url.endsWith("/src/data/icons.json")) return jsonResponse(registries["thesvg"]);
    if (url.endsWith("/index.json")) return jsonResponse(registries["selfhst"]);
    if (url.endsWith("/metadata.json")) return jsonResponse(registries["dashboardMetadata"]);
    if (url.endsWith("/tree.json")) return jsonResponse(registries["dashboardTree"]);
    return new Response("not found", { status: 404 });
  }));
}

function registryFixtures(): Record<string, unknown> {
  return {
    thesvg: [
      {
        slug: "cf-registry",
        title: "CF Registry",
        variants: { default: "/icons/cf-registry/default.svg" },
      },
    ],
    selfhst: [
      {
        Reference: "cf-selfhst",
        Name: "CF selfh.st",
        SVG: "Yes",
      },
    ],
    dashboardMetadata: {
      "cf-dashboard": {},
    },
    dashboardTree: {
      svg: ["cf-dashboard.svg"],
    },
  };
}

function providerVersion(commitSha: string) {
  return {
    sourceRef: commitSha,
    displayVersion: commitSha.slice(0, 7),
    commitSha,
    commitShortSha: commitSha.slice(0, 7),
    commitDate: "2026-06-11T00:00:00Z",
    releaseTag: null,
    releasePublishedAt: null,
  };
}

function jsonResponse(value: unknown, headers: Record<string, string> = {}): Response {
  return new Response(JSON.stringify(value), {
    headers: { "content-type": "application/json", ...headers },
  });
}

function stringArg(args: readonly unknown[], index: number): string {
  const value = args[index];
  return typeof value === "string" ? value : "";
}

function numberArg(args: readonly unknown[], index: number): number {
  const value = args[index];
  return typeof value === "number" ? value : 0;
}
